packages/edit-widgets/src/blocks/legacy-widget/index.js

```diff
@@ -34,49 +40,74 @@ export const settings = {
  * @return {Object} Block object.
  */
 export const create = ( editorSettings ) => {
-	const legacyWidgets = editorSettings?.availableLegacyWidgets ?? {};
+	const legacyWidgets = Object.entries(
+		editorSettings.availableLegacyWidgets
+	).filter( ( [ , widget ] ) => ! widget.isHidden );
+
+	return [
+		{
+			...blockBasis,
+			settings: {
+				...blockBasis.settings,
+				variations: legacyWidgets
+					.filter( ( [ , widget ] ) => ! widget.isReferenceWidget )
+					.map( ( [ className, widget ] ) =>
+						legacyWidgetToBlockVariation( className, widget )
+					),
+			},
+		},
+		...legacyWidgets
+			.filter( ( [ , widget ] ) => widget.isReferenceWidget )
+			.map( ( [ className, widget ] ) =>
+				referenceLegacyWidgetsToBlockType( className, widget )
+			),
+	];
+};
+
+function referenceLegacyWidgetsToBlockType( className, widget ) {
+	const attrs = blockBasis.metadata.attributes;
 	return {
-		metadata,
-		name,
+		...blockBasis,
+		name: widget.blockName,
+		metadata: {
+			...blockBasis.metadata,
+			name: widget.blockName,
+			supports: {
+				...blockBasis.metadata.supports,
+				multiple: false,
+			},
+			attributes: {
+				...attrs,
+				referenceWidgetName: {
+					...attrs.referenceWidgetName,
+					default: className,
+				},
+				instance: {
+					...attrs.instance,
+					default: {},
+				},
+			},
+		},
```

Comment:
Maybe it's time we bring `immer` into our toolchains 😅 

Or before that happens, could we just `deepClone` the object and mutate it here?